# Battle for the Neighborhoods Final: Paw & Claw

IBM Data Science Capstone | Final Class Assignment | Student: DL | 02.07.2021 |

*Paw & Claw* (our fictional business) is a pet sitting, walking, grooming and daycare service seeking to expand in the city of Portland, OR.

## The Problem to Solve

What is the ideal location intersecting near pet supply stores, and parks, with current limited existing Pet Services?

I will examine Apartments & Residential Buildings, Pet services and stores, Trails, and Grocery stores.  I will then examine area of Portland with the highest density of apartments and find the area closest to Trails, with some pet services and grocery stores to make a recommendation for a new location.

Requirement:

- *Paw & Claw* report that the majority of customers that use their services live in apartments or condos and only will consider a new location near Apartment residential building locations.

## Criteria

Do residents in the area travel frequently?
Paw & Claw currently partners with local travel agencies by placing kiosks and brochures in agency offices and have a marketing process set up for turn key integration with travel agency software and sales.
-  *Paw & Claw* has requested to explore the proposed new location to determine if it would be suitable to include an expanded boarding service.  

## Methodology

- Use Visualization analysis using Foursquare data to answer *Paw & Claw's* questions

- Explore Travel agencies in within a 1 mile radius of the proposed new location and provide a list of agency names and addresses, and my recommendations.

## Data sources

Foursquare API venues, categories: Apartments, Pet, Grocery, Trail, Travel [Link to Foursquare Categories](https://developer.foursquare.com/docs/build-with-foursquare/categories/)


# Importing libraries

In [418]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Libraries imported.


Credentials and Geolocation foursquare_agent

In [419]:
# The code was removed by Watson Studio for sharing.

In [420]:
address = 'Portland, OR'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)

45.5202471 -122.6741949


Searching Foursquare: Apartment, with categoryId, 2 mile radius of downtown Porland

In [421]:
search_query = 'Apartment'
categoryId = '4d954b06a243a5684965b473'
radius = 1000
print(search_query + ' .... OK!')

Apartment .... OK!


Get URL

In [422]:
# The code was removed by Watson Studio for sharing.

In [423]:
apartments = requests.get(url).json()

In [424]:
# assign relevant part of JSON to venues
venues = apartments['response']['venues']
# tranform venues into a dataframe
apartments_dataframe = json_normalize(venues)
apartments_dataframe.head(50)

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:4: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  after removing the cwd from sys.path.


id                                      name  \
0   5878842c7a8f042d8865a59b  Bar Bar Apartment at Mississippi Studios   
1   5c168fcf4f0e58002b25db72            10Th@Hoyt Apartment Think Tank   
2   4d5b01117c3b224b4e2823af                    Essex House Apartments   
3   4bf0c96917880f47dab72937                   Pacifc Tower Apartments   
4   5c059720180b91002c87b767                       38 Davis Apartments   
5   553f27b6498e9f38c2cd707c                    The Jeffrey Apartments   
6   4acd3848f964a52075cb20e3                       Cyan PDX Apartments   
7   4bfb5364d0382d7f4c97c90a                        Ongford Apartments   
8   4d43abee1928a35d2801b170       5th Ave Place Apartments -silverado   
9   577090c1498e3df0128ed6cc                      Alder St. Apartments   
10  56f9b242498edbb3493bf705                            mw8 apartments   
11  5882a269916802691bb6288a                     11 Marche' Apartments   
12  4f03b404823162ada60fc999                     Ambassador Apartments   
13  55c6d1e1498e06fe7ce8579f                  The Socereign Apartments   
14  4bc78aad14d79521599567e9                         Indigo Apartments   
15  5824dc0cd506b731bd0711b8                        Regency Apartments   
16  5d28cc2038c7d60023c125aa                     The Louisa Apartments   
17  4bc250b2abf495218f60c293                    10th @ Hoyt Apartments   
18  5780f34ecd1066b31e3b4f23                West Park Place Apartments   
19  4d0a94bbb93b224b111247bf                       Clifford Apartments   
20  50494c41e4b0f680c83da179                      The Janey Apartments   
21  4c2840f5fe6e2d7f1509533c                          Brown Apartments   
22  4b118891f964a520ac7e23e3                  The Sovereign Apartments   
23  4e30a55c62e11193f6e12077                      Pine Cone Apartments   

                                           categories    referralId  hasPerk  \
0   [{'id': '4bf58dd8d48988d1e5931735', 'name': 'M...  v-1612779215    False   
1   [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
2   [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
3   [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
4   [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
5   [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
6   [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
7   [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
8   [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
9   [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
10  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
11  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
12  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
13  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
14  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
15  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
16  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
17  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
18  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
19  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
20  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
21  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
22  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   
23  [{'id': '4d954b06a243a5684965b473', 'name': 'R...  v-1612779215    False   

    location.lat  location.lng  \
0      45.523450   -122.676210   
1      45.527333   -122.680923   
2      45.513590   -122.677571   
3      45.525988   -122.674873   
4   

Cleaning

In [425]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in apartments_dataframe.columns if col.startswith('location.')] + ['id']
apartments_dataframe_filtered = apartments_dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
apartments_dataframe_filtered['categories'] = apartments_dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
apartments_dataframe_filtered.columns = [column.split('.')[-1] for column in apartments_dataframe_filtered.columns]

apartments_dataframe_filtered.head()

name  \
0  Bar Bar Apartment at Mississippi Studios   
1            10Th@Hoyt Apartment Think Tank   
2                    Essex House Apartments   
3                   Pacifc Tower Apartments   
4                       38 Davis Apartments   

                                 categories        lat         lng  \
0                               Music Venue  45.523450 -122.676210   
1  Residential Building (Apartment / Condo)  45.527333 -122.680923   
2  Residential Building (Apartment / Condo)  45.513590 -122.677571   
3  Residential Building (Apartment / Condo)  45.525988 -122.674873   
4  Residential Building (Apartment / Condo)  45.524309 -122.670799   

                                      labeledLatLngs  distance  cc      city  \
0  [{'label': 'display', 'lat': 45.52345, 'lng': ...       389  US  Portland   
1  [{'label': 'display', 'lat': 45.527333, 'lng':...       947  US  Portland   
2  [{'label': 'display', 'lat': 45.51358952456643...       786  US  Portland   
3  [{'label': 'display', 'lat': 45.52598756506867...       641  US  Portland   
4  [{'label': 'display', 'lat': 45.524309, 'lng':...       524  US  Portland   

  state        country                                 formattedAddress  \
0    OR  United States                                   [Portland, OR]   
1    OR  United States                             [Portland, OR 97209]   
2    OR  United States     [1330 Southwest 3rd Ave, Portland, OR 97201]   
3    OR  United States  [333 NW 4th Ave (Flanders), Portland, OR 97209]   
4    OR  United States             [38 NW Davis St, Portland, OR 97209]   

  postalCode                 address        neighborhood crossStreet  \
0        NaN                     NaN                 NaN         NaN   
1      97209                     NaN                 NaN         NaN   
2      97201  1330 Southwest 3rd Ave   Downtown Portland         NaN   
3      97209          333 NW 4th Ave                 NaN    Flanders   
4      97209          38 NW Davis St  Old Town Chinatown         NaN   

                         id  
0  5878842c7a8f042d8865a59b  
1  5c168fcf4f0e58002b25db72  
2  4d5b01117c3b224b4e2823af  
3  4bf0c96917880f47dab72937  
4  5c059720180b91002c87b767

Foursquare returned accurate results, all categories are related to apartments.

In [426]:
apartments_dataframe_filtered.name

0     Bar Bar Apartment at Mississippi Studios
1               10Th@Hoyt Apartment Think Tank
2                       Essex House Apartments
3                      Pacifc Tower Apartments
4                          38 Davis Apartments
5                       The Jeffrey Apartments
6                          Cyan PDX Apartments
7                           Ongford Apartments
8          5th Ave Place Apartments -silverado
9                         Alder St. Apartments
10                              mw8 apartments
11                       11 Marche' Apartments
12                       Ambassador Apartments
13                    The Socereign Apartments
14                           Indigo Apartments
15                          Regency Apartments
16                       The Louisa Apartments
17                      10th @ Hoyt Apartments
18                  West Park Place Apartments
19                         Clifford Apartments
20                        The Janey Apartments
21           

# Pet related services and retail
Convenience stores are included as they sell pet food

In [427]:
search_query = 'Pet'
categoryId = '5032897c91d4c4b30a586d69'
radius = 1800
print(search_query + ' .... OK!')

Pet .... OK!


Corresponding URL

In [428]:
# The code was removed by Watson Studio for sharing.

In [429]:
pet = requests.get(url).json()

In [430]:
# assign relevant part of JSON to venues
venues = pet['response']['venues']

# tranform venues into a dataframe
pet_dataframe = json_normalize(venues)
pet_dataframe.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  """


id                             name  \
0  4b84629ff964a520543231e3  Nature's Pet Market NW Portland   
1  52ae4a76498e10465a9a291a            Banfield Pet Hospital   
2  4f322c1619836c91c7bd4715             Petite Cafe and Deli   
3  5b7dca4212138400394c2274            Banfield Pet Hospital   
4  55ea217b498e5acd4c97e271              RiverPlace Pet Park   

                                          categories    referralId  hasPerk  \
0  [{'id': '4bf58dd8d48988d100951735', 'name': 'P...  v-1612779220    False   
1  [{'id': '4d954af4a243a5684765b473', 'name': 'V...  v-1612779220    False   
2  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...  v-1612779220    False   
3  [{'id': '4d954af4a243a5684765b473', 'name': 'V...  v-1612779220    False   
4  [{'id': '4bf58dd8d48988d1e5941735', 'name': 'D...  v-1612779220    False   

        location.address location.crossStreet  location.lat  location.lng  \
0        111 NW 21st Ave             Burnside     45.523958   -122.694441   
1          35 NW 20th Pl                  NaN     45.523750   -122.693631   
2   319 SW Washington St                  NaN     45.519878   -122.675034   
3  940 SE 11th Ave Ste C                  NaN     45.516245   -122.654551   
4           River Marina         River Circle     45.509280   -122.674819   

                             location.labeledLatLngs  location.distance  \
0  [{'label': 'display', 'lat': 45.52395815436707...               1632   
1  [{'label': 'display', 'lat': 45.52375003959196...               1565   
2  [{'label': 'display', 'lat': 45.51987838745117...                 77   
3  [{'label': 'entrance', 'lat': 45.516042, 'lng'...               1595   
4  [{'label': 'display', 'lat': 45.50928030960593...               1221   

  location.postalCode location.cc location.city location.state  \
0               97209          US      Portland             OR   
1               97209          US      Portland             OR   
2               97204          US      Portland             OR   
3               97214          US      Portland             OR   
4               97201          US      Portland             OR   

  location.country                          location.formattedAddress  \
0    United States   [111 NW 21st Ave (Burnside), Portland, OR 97209]   
1    United States                [35 NW 20th Pl, Portland, OR 97209]   
2    United States         [319 SW Washington St, Portland, OR 97204]   
3    United States        [940 SE 11th Ave Ste C, Portland, OR 97214]   
4    United States  [River Marina (River Circle), Portland, OR 97201]   

  venuePage.id delivery.id delivery.url delivery.provider.name  \
0          NaN         NaN          NaN                    NaN   
1          NaN         NaN          NaN                    NaN   
2          NaN         NaN          NaN                    NaN   
3          NaN         NaN          NaN                    NaN   
4          NaN         NaN          NaN                    NaN   

  delivery.provider.icon.prefix delivery.provider.icon.sizes  \
0                           NaN                          NaN   
1                           NaN                          NaN   
2                           NaN                          NaN   
3                           NaN                          NaN   
4                           NaN                          NaN   

  delivery.provider.icon.name  
0                         NaN  
1                         NaN  
2                         NaN  
3                         NaN  
4                         NaN

Cleaning

In [431]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in pet_dataframe.columns if col.startswith('location.')] + ['id']
pet_dataframe_filtered = pet_dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
pet_dataframe_filtered['categories'] = pet_dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
pet_dataframe_filtered.columns = [column.split('.')[-1] for column in pet_dataframe_filtered.columns]

pet_dataframe_filtered.head(20)

name                  categories  \
0       Nature's Pet Market NW Portland                   Pet Store   
1                 Banfield Pet Hospital                Veterinarian   
2                  Petite Cafe and Deli                 Coffee Shop   
3                 Banfield Pet Hospital                Veterinarian   
4                   RiverPlace Pet Park                     Dog Run   
5                            Peterson's           Convenience Store   
6                     peter max exhibit                 Art Gallery   
7     Peter R. Martin, Balance Medicine                         Spa   
8    The Law Offices of Peter O. Hansen                      Lawyer   
9        Law Offices of Peter O. Hansen  Financial or Legal Service   
10                           Peterson's               Grocery Store   
11              PETTY JILL-DUI Attorney                      Lawyer   
12                      peter evans Law                      Lawyer   
13                          Petit Panda            Asian Restaurant   
14                   Peterson's Grocery           Convenience Store   
15                  Bistro Petit Oiseau           French Restaurant   
16             Rose Petal Massage & Spa              Massage Studio   
17                  Jill Petty Law Firm                      Lawyer   
18                     Papa Petes Pizza                 Pizza Place   
19  Peter Park, J.D., Real Estate Agent          Real Estate Office   

                        address     crossStreet        lat         lng  \
0               111 NW 21st Ave        Burnside  45.523958 -122.694441   
1                 35 NW 20th Pl             NaN  45.523750 -122.693631   
2          319 SW Washington St             NaN  45.519878 -122.675034   
3         940 SE 11th Ave Ste C             NaN  45.516245 -122.654551   
4                  River Marina    River Circle  45.509280 -122.674819   
5                513 SW 4th Ave             NaN  45.519675 -122.676001   
6                           NaN             NaN  45.518131 -122.676450   
7   1130 SW Morrison St Ste 430             NaN  45.519360 -122.679028   
8       620 SW 5th Ave Ste 1210             NaN  45.519180 -122.677002   
9           620 SW 5th Ave #604             NaN  45.519169 -122.677002   
10            115 SW Yamhill St             NaN  45.517553 -122.674049   
11              1001 SW 5th Ave             NaN  45.516610 -122.679360   
12                 308 Sw First       Oak @ 1st  45.525890 -122.673775   
13                   Se 9th Ave   Sw Washington  45.520896 -122.680621   
14             1143 SW Stark St  SW 12th Avenue  45.522667 -122.682760   
15                          NaN             NaN  45.521909 -122.678023   
16                          NaN             NaN  45.518214 -122.676552   
17              1001 SW 5th Ave             NaN  45.516572 -122.679332   
18               Pioneer street             NaN  45.514317 -122.680767   
19           650 NE Holladay St             NaN  45.529764 -122.659151   

                                       labeledLatLngs  distance postalCode  \
0   [{'label': 'display', 'lat': 45.52395815436707...      1632      97209   
1   [{'label': 'display', 'lat': 45.52375003959196...      1565      97209   
2   [{'label': 'display', 'lat': 45.51987838745117...        77      97204   
3   [{'label': 'entrance', 'lat': 45.516042, 'lng'...      1595      97214   
4   [{'label': 'display', 'lat': 45.50928030960593...      1221      97201   
5   [{'label': 'display', 'lat': 45.51967482105092...       154      97204   
6   [{'label': 'display', 'lat': 45.51813136148788...       293        NaN   
7   [{'label': 'display', 'lat': 45.51936, 'lng': ...       389      97205   
8   [{'label': 'display', 'lat': 45.5191802978516,...       249      97204   
9   [{'label': 'display', 'lat': 45.51916885375976...       249      97204   
10  [{'label': 'display', 'lat': 45.517553, 'lng':...       300      97204   
11  [{'label': 'display', 'lat': 45.51661, 'lng': ..

# Cleaning: Pet dataframe, dropping non-related categories returned by Foursquare
I know this is not very elegant, but I could not remember how to do this in one line in particular inplace=True, so performed the operation seperately

In [432]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='Coffee Shop'].index, inplace=True)

In [433]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='Art Gallery'].index, inplace=True)

In [434]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='Spa'].index, inplace=True)

In [435]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='Bakery'].index, inplace=True)

In [436]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='Massage Studio'].index, inplace=True)

In [437]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='Grocery Store'].index, inplace=True)

In [438]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='Lawyer'].index, inplace=True)

In [439]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='Asian Restaurant'].index, inplace=True)

In [440]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='Financial or Legal Service'].index, inplace=True)

In [441]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='French Restaurant'].index, inplace=True)

In [442]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='College Rec Center'].index, inplace=True)

In [443]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='Pizza Place'].index, inplace=True)

In [444]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='Real Estate Office'].index, inplace=True)

In [445]:
pet_dataframe_filtered.drop(pet_dataframe_filtered.loc[pet_dataframe_filtered['categories']=='College Tennis Court'].index, inplace=True)

Dropping unecessary columns

In [446]:
pet_dataframe_filtered.drop(["crossStreet", "cc", "city", "state","country"], axis=1, inplace = True)

In [447]:
pet_dataframe_filtered.head(50)

name                                categories  \
0   Nature's Pet Market NW Portland                                 Pet Store   
1             Banfield Pet Hospital                              Veterinarian   
3             Banfield Pet Hospital                              Veterinarian   
4               RiverPlace Pet Park                                   Dog Run   
5                        Peterson's                         Convenience Store   
14               Peterson's Grocery                         Convenience Store   
20             Peter Niland.com LLC                          Business Service   
21                       Peterson's                         Convenience Store   
22               Pettygrove Station                  Mediterranean Restaurant   
25      Peter Corvallis Productions                            Rental Service   
26           1942 NW Pettygrove st,  Residential Building (Apartment / Condo)   
27   Dr Peter leitner- chiropractor                              Chiropractor   
28                  Pettygrove Park                                      Park   
29      Pettygrove Physical Therapy                        Physical Therapist   
30                       Noah's Arf                                 Pet Store   

                  address        lat         lng  \
0         111 NW 21st Ave  45.523958 -122.694441   
1           35 NW 20th Pl  45.523750 -122.693631   
3   940 SE 11th Ave Ste C  45.516245 -122.654551   
4            River Marina  45.509280 -122.674819   
5          513 SW 4th Ave  45.519675 -122.676001   
14       1143 SW Stark St  45.522667 -122.682760   
20      2014 NW Glisan St  45.525966 -122.692770   
21     922 SW Morrison St  45.519963 -122.681710   
22       1401 NW 19th Ave  45.532936 -122.690918   
25  2200 N Interstate Ave  45.538457 -122.673613   
26                    NaN  45.529017 -122.677117   
27        522 NW 12th Ave  45.526940 -122.682981   
28        1800 SW 1st Ave  45.510990 -122.678697   
29                    NaN  45.533738 -122.689748   
30       1306 NW 18th Ave  45.532518 -122.689182   

                                       labeledLatLngs  distance postalCode  \
0   [{'label': 'display', 'lat': 45.52395815436707...      1632      97209   
1   [{'label': 'display', 'lat': 45.52375003959196...      1565      97209   
3   [{'label': 'entrance', 'lat': 45.516042, 'lng'...      1595      97214   
4   [{'label': 'display', 'lat': 45.50928030960593...      1221      97201   
5   [{'label': 'display', 'lat': 45.51967482105092...       154      97204   
14  [{'label': 'display', 'lat': 45.522667, 'lng':...       720      97205   
20  [{'label': 'display', 'lat': 45.525966, 'lng':...      1582      97209   
21  [{'label': 'display', 'lat': 45.51996313884486...       586      97205   
22  [{'label': 'entrance', 'lat': 45.532946, 'lng'...      1922      97209   
25  [{'label': 'display', 'lat': 45.538457, 'lng':...      2027      97227   
26  [{'label': 'display', 'lat': 45.529017, 'lng':...      1002      97209   
27  [{'label': 'display', 'lat': 45.52694, 'lng': ...      1012      97209   
28  [{'label': 'display', 'lat': 45.51098987235108...      1088      97201   
29  [{'label': 'display', 'lat': 45.53373764719377...      1930      97209   
30  [{'label': 'display', 'lat': 45.53251799999999...      1797      97209   

                                     formattedAddress  \
0    [111 NW 21st Ave (Burnside), Portland, OR 97209]   
1                 [35 NW 20th Pl, Portland, OR 97209]   
3         [940 SE 11th Ave Ste C, Portland, OR 97214]   
4   [River Marina (River Circle), Portland, OR 97201]   
5                [513 SW 4th Ave, Portland, OR 97204]   
14  [1143 SW Stark St (SW 12th Avenue), Portland, ...   
20            [2014 NW Glisan St, Portland, OR 97209]   
21  [922 SW Morrison St (at SW 9th Ave), Portland,...   
22             [1401 NW 19th Ave, Portland, OR 97209]   
25  [2200 N Interstate Ave (Thompson), Portland, O...   
26                       

In [448]:
pet_dataframe_filtered.name

0     Nature's Pet Market NW Portland
1               Banfield Pet Hospital
3               Banfield Pet Hospital
4                 RiverPlace Pet Park
5                          Peterson's
14                 Peterson's Grocery
20               Peter Niland.com LLC
21                         Peterson's
22                 Pettygrove Station
25        Peter Corvallis Productions
26             1942 NW Pettygrove st,
27     Dr Peter leitner- chiropractor
28                    Pettygrove Park
29        Pettygrove Physical Therapy
30                         Noah's Arf
Name: name, dtype: object

# Grocery
Where do pet owners possibly buy their pet food? I want to see if there are nearby grocery stores in addition to the 2 pet stores, within 25 mile radius

In [449]:
search_query = 'Grocery'
categoryId = '4bf58dd8d48988d1f9941735'
radius = 1000
print(search_query + ' .... OK!')

Grocery .... OK!


Corresponding URL

In [450]:
# The code was removed by Watson Studio for sharing.

In [451]:
grocery = requests.get(url).json()

In [452]:
# assign relevant part of JSON to venues
venues = grocery['response']['venues']

# tranform venues into a dataframe
grocery_dataframe = json_normalize(venues)
grocery_dataframe.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  """


id                      name  \
0  4c411003af052d7f0adf7c79  Katinas Deli And Grocery   
1  4dee3fcae4cd032b189d7cbf                Tk Grocery   
2  4b8553d2f964a520ec5631e3      T & K Grocery & Deli   
3  4fa1797ee4b0dae72e4aad66         Junior II Grocery   
4  4a9f16eef964a520793c20e3          Old Town Grocery   

                                          categories    referralId  hasPerk  \
0  [{'id': '4bf58dd8d48988d146941735', 'name': 'D...  v-1612779232    False   
1  [{'id': '4d954b0ea243a5684a65b473', 'name': 'C...  v-1612779232    False   
2  [{'id': '4bf58dd8d48988d118951735', 'name': 'G...  v-1612779232    False   
3  [{'id': '4d954b0ea243a5684a65b473', 'name': 'C...  v-1612779232    False   
4  [{'id': '4bf58dd8d48988d118951735', 'name': 'G...  v-1612779232    False   

     location.address location.crossStreet  location.lat  location.lng  \
0  612 NW Flanders St              6th ave     45.525701   -122.676847   
1                 NaN                  NaN     45.520347   -122.674177   
2      332 SW 3rd Ave            SW Oak St     45.520496   -122.674123   
3                 NaN                  NaN     45.522345   -122.677994   
4      100 NW 3rd Ave          NW Couch St     45.524068   -122.673104   

                             location.labeledLatLngs  location.distance  \
0  [{'label': 'display', 'lat': 45.525701, 'lng':...                641   
1  [{'label': 'display', 'lat': 45.52034743223867...                 11   
2  [{'label': 'display', 'lat': 45.5204960551861,...                 28   
3  [{'label': 'display', 'lat': 45.52234525982983...                377   
4  [{'label': 'display', 'lat': 45.524068, 'lng':...                433   

  location.postalCode location.cc location.city location.state  \
0               97209          US      Portland             OR   
1                 NaN          US      Portland             OR   
2               97204          US      Portland             OR   
3               97205          US      Portland             OR   
4               97209          US      Portland             OR   

  location.country                          location.formattedAddress  \
0    United States  [612 NW Flanders St (6th ave), Portland, OR 97...   
1    United States                                     [Portland, OR]   
2    United States   [332 SW 3rd Ave (SW Oak St), Portland, OR 97204]   
3    United States                               [Portland, OR 97205]   
4    United States  [100 NW 3rd Ave (NW Couch St), Portland, OR 97...   

  location.neighborhood  
0                   NaN  
1                   NaN  
2                   NaN  
3                   NaN  
4                   NaN

In [453]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in grocery_dataframe.columns if col.startswith('location.')] + ['id']
grocery_dataframe_filtered = grocery_dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
grocery_dataframe_filtered['categories'] = grocery_dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
grocery_dataframe_filtered.columns = [column.split('.')[-1] for column in grocery_dataframe_filtered.columns]

grocery_dataframe_filtered.head(5)

name         categories             address  \
0  Katinas Deli And Grocery      Deli / Bodega  612 NW Flanders St   
1                Tk Grocery  Convenience Store                 NaN   
2      T & K Grocery & Deli      Grocery Store      332 SW 3rd Ave   
3         Junior II Grocery  Convenience Store                 NaN   
4          Old Town Grocery      Grocery Store      100 NW 3rd Ave   

   crossStreet        lat         lng  \
0      6th ave  45.525701 -122.676847   
1          NaN  45.520347 -122.674177   
2    SW Oak St  45.520496 -122.674123   
3          NaN  45.522345 -122.677994   
4  NW Couch St  45.524068 -122.673104   

                                      labeledLatLngs  distance postalCode  cc  \
0  [{'label': 'display', 'lat': 45.525701, 'lng':...       641      97209  US   
1  [{'label': 'display', 'lat': 45.52034743223867...        11        NaN  US   
2  [{'label': 'display', 'lat': 45.5204960551861,...        28      97204  US   
3  [{'label': 'display', 'lat': 45.52234525982983...       377      97205  US   
4  [{'label': 'display', 'lat': 45.524068, 'lng':...       433      97209  US   

       city state        country  \
0  Portland    OR  United States   
1  Portland    OR  United States   
2  Portland    OR  United States   
3  Portland    OR  United States   
4  Portland    OR  United States   

                                    formattedAddress neighborhood  \
0  [612 NW Flanders St (6th ave), Portland, OR 97...          NaN   
1                                     [Portland, OR]          NaN   
2   [332 SW 3rd Ave (SW Oak St), Portland, OR 97204]          NaN   
3                               [Portland, OR 97205]          NaN   
4  [100 NW 3rd Ave (NW Couch St), Portland, OR 97...          NaN   

                         id  
0  4c411003af052d7f0adf7c79  
1  4dee3fcae4cd032b189d7cbf  
2  4b8553d2f964a520ec5631e3  
3  4fa1797ee4b0dae72e4aad66  
4  4a9f16eef964a520793c20e3

In [454]:
grocery_dataframe_filtered.drop(["crossStreet", "cc", "city", "state","country","neighborhood"], axis=1, inplace = True)

In [455]:
grocery_dataframe_filtered.head()

name         categories             address        lat  \
0  Katinas Deli And Grocery      Deli / Bodega  612 NW Flanders St  45.525701   
1                Tk Grocery  Convenience Store                 NaN  45.520347   
2      T & K Grocery & Deli      Grocery Store      332 SW 3rd Ave  45.520496   
3         Junior II Grocery  Convenience Store                 NaN  45.522345   
4          Old Town Grocery      Grocery Store      100 NW 3rd Ave  45.524068   

          lng                                     labeledLatLngs  distance  \
0 -122.676847  [{'label': 'display', 'lat': 45.525701, 'lng':...       641   
1 -122.674177  [{'label': 'display', 'lat': 45.52034743223867...        11   
2 -122.674123  [{'label': 'display', 'lat': 45.5204960551861,...        28   
3 -122.677994  [{'label': 'display', 'lat': 45.52234525982983...       377   
4 -122.673104  [{'label': 'display', 'lat': 45.524068, 'lng':...       433   

  postalCode                                   formattedAddress  \
0      97209  [612 NW Flanders St (6th ave), Portland, OR 97...   
1        NaN                                     [Portland, OR]   
2      97204   [332 SW 3rd Ave (SW Oak St), Portland, OR 97204]   
3      97205                               [Portland, OR 97205]   
4      97209  [100 NW 3rd Ave (NW Couch St), Portland, OR 97...   

                         id  
0  4c411003af052d7f0adf7c79  
1  4dee3fcae4cd032b189d7cbf  
2  4b8553d2f964a520ec5631e3  
3  4fa1797ee4b0dae72e4aad66  
4  4a9f16eef964a520793c20e3

In [456]:
grocery_dataframe_filtered.name

0       Katinas Deli And Grocery
1                     Tk Grocery
2           T & K Grocery & Deli
3              Junior II Grocery
4               Old Town Grocery
5              My Street Grocery
6        Georgia's Grocery Store
7              Georgia's Grocery
8             Peterson's Grocery
9            Day & Night Grocery
10             West Side Grocery
11           Green Zebra Grocery
12    Sixth Ave Grocery and Deli
13              4th & Oak Market
Name: name, dtype: object

# Trails
I want to find trails nearby ideal for pet walking and pet day sitting activity. I use the Outdoors & Recreation main categoryId and Trails as my search_query

In [457]:
search_query = 'Trail'
categoryId = '4bf58dd8d48988d159941735'
radius = 2000
print(search_query + ' .... OK!')

Trail .... OK!


Defining URL

In [458]:
# The code was removed by Watson Studio for sharing.

GET response

In [459]:
trail= requests.get(url).json()

Put into pandas dataframe

In [460]:
# assign relevant part of JSON to venues
venues = trail['response']['venues']

# tranform venues into a dataframe
trail_dataframe = json_normalize(venues)
trail_dataframe.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  """


id                                 name  \
0  4cc8a97041e754818ff25784             Forest Park - Keil Trail   
1  4b4e4b34f964a52062e826e3  Portland Trail Blazers Front Office   
2  5a79fb5b6336be4255cae676                Oregon Trail Building   
3  5532ac77498eca93a67815c2            Forest Park - Maple Trail   
4  53ffa848498e28387bdcab7d             Trail Blazers Digital HQ   

                                          categories    referralId  hasPerk  \
0  [{'id': '4bf58dd8d48988d159941735', 'name': 'T...  v-1612779236    False   
1  [{'id': '4bf58dd8d48988d124941735', 'name': 'O...  v-1612779236    False   
2  [{'id': '4bf58dd8d48988d130941735', 'name': 'B...  v-1612779236    False   
3  [{'id': '52e81612bcbc57f1066b7a21', 'name': 'N...  v-1612779236    False   
4  [{'id': '52e81612bcbc57f1066b7a33', 'name': 'S...  v-1612779236    False   

      location.address  location.lat  location.lng  \
0         NW 53rd Ave.     45.529571   -122.685720   
1  1 N Center Court St     45.531861   -122.667751   
2       333 SW 5th Ave     45.520974   -122.676708   
3                  NaN     45.522940   -122.682408   
4  1 N Center Court St     45.531517   -122.667686   

                             location.labeledLatLngs  location.distance  \
0  [{'label': 'display', 'lat': 45.52957082377107...               1373   
1  [{'label': 'display', 'lat': 45.53186050825756...               1387   
2  [{'label': 'display', 'lat': 45.520974, 'lng':...                212   
3  [{'label': 'display', 'lat': 45.52294036157403...                707   
4  [{'label': 'display', 'lat': 45.53151702880859...               1353   

  location.postalCode location.cc location.city location.state  \
0               97210          US      Portland             OR   
1               97227          US      Portland             OR   
2               97204          US      Portland             OR   
3                 NaN          US      Portland             OR   
4               97227          US      Portland             OR   

  location.country                  location.formattedAddress  \
0    United States         [NW 53rd Ave., Portland, OR 97210]   
1    United States  [1 N Center Court St, Portland, OR 97227]   
2    United States       [333 SW 5th Ave, Portland, OR 97204]   
3    United States                             [Portland, OR]   
4    United States  [1 N Center Court St, Portland, OR 97227]   

  location.crossStreet  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN

In [461]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in trail_dataframe.columns if col.startswith('location.')] + ['id']
trail_dataframe_filtered = trail_dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
trail_dataframe_filtered['categories'] = trail_dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
trail_dataframe_filtered.columns = [column.split('.')[-1] for column in trail_dataframe_filtered.columns]

trail_dataframe_filtered.head()

name     categories              address  \
0             Forest Park - Keil Trail          Trail         NW 53rd Ave.   
1  Portland Trail Blazers Front Office         Office  1 N Center Court St   
2                Oregon Trail Building       Building       333 SW 5th Ave   
3            Forest Park - Maple Trail  National Park                  NaN   
4             Trail Blazers Digital HQ    Social Club  1 N Center Court St   

         lat         lng                                     labeledLatLngs  \
0  45.529571 -122.685720  [{'label': 'display', 'lat': 45.52957082377107...   
1  45.531861 -122.667751  [{'label': 'display', 'lat': 45.53186050825756...   
2  45.520974 -122.676708  [{'label': 'display', 'lat': 45.520974, 'lng':...   
3  45.522940 -122.682408  [{'label': 'display', 'lat': 45.52294036157403...   
4  45.531517 -122.667686  [{'label': 'display', 'lat': 45.53151702880859...   

   distance postalCode  cc      city state        country  \
0      1373      97210  US  Portland    OR  United States   
1      1387      97227  US  Portland    OR  United States   
2       212      97204  US  Portland    OR  United States   
3       707        NaN  US  Portland    OR  United States   
4      1353      97227  US  Portland    OR  United States   

                            formattedAddress crossStreet  \
0         [NW 53rd Ave., Portland, OR 97210]         NaN   
1  [1 N Center Court St, Portland, OR 97227]         NaN   
2       [333 SW 5th Ave, Portland, OR 97204]         NaN   
3                             [Portland, OR]         NaN   
4  [1 N Center Court St, Portland, OR 97227]         NaN   

                         id  
0  4cc8a97041e754818ff25784  
1  4b4e4b34f964a52062e826e3  
2  5a79fb5b6336be4255cae676  
3  5532ac77498eca93a67815c2  
4  53ffa848498e28387bdcab7d

# Cleaning: Dropping values in Rows that are not trails. 

In [462]:
trail_dataframe_filtered.drop(trail_dataframe_filtered.loc[trail_dataframe_filtered['categories']=='Office'].index, inplace=True)

In [463]:
trail_dataframe_filtered.drop(trail_dataframe_filtered.loc[trail_dataframe_filtered['categories']=='Building'].index, inplace=True)

In [464]:
trail_dataframe_filtered.drop(trail_dataframe_filtered.loc[trail_dataframe_filtered['categories']=='Credit Union'].index, inplace=True)

In [465]:
trail_dataframe_filtered.drop(trail_dataframe_filtered.loc[trail_dataframe_filtered['categories']=='Social Club'].index, inplace=True)

In [466]:
trail_dataframe_filtered.drop(trail_dataframe_filtered.loc[trail_dataframe_filtered['categories']=='Brewery'].index, inplace=True)

In [467]:
trail_dataframe_filtered.drop(trail_dataframe_filtered.loc[trail_dataframe_filtered['categories']=='Clothing Store'].index, inplace=True)

In [468]:
trail_dataframe_filtered.drop(["crossStreet", "cc", "city", "state","country","crossStreet"], axis=1, inplace = True)

In [469]:
trail_dataframe_filtered.head()

name        categories       address        lat  \
0         Forest Park - Keil Trail             Trail  NW 53rd Ave.  45.529571   
3        Forest Park - Maple Trail     National Park           NaN  45.522940   
5  Willamette River Greenway Trail             Trail           NaN  45.529514   
6              SW Pedestrian Trail  Pedestrian Plaza           NaN  45.510440   
7                   Hillside Trail             Trail           NaN  45.512758   

          lng                                     labeledLatLngs  distance  \
0 -122.685720  [{'label': 'display', 'lat': 45.52957082377107...      1373   
3 -122.682408  [{'label': 'display', 'lat': 45.52294036157403...       707   
5 -122.673455  [{'label': 'display', 'lat': 45.529514, 'lng':...      1033   
6 -122.679800  [{'label': 'display', 'lat': 45.51044, 'lng': ...      1176   
7 -122.688448  [{'label': 'display', 'lat': 45.51275759151022...      1389   

  postalCode                    formattedAddress                        id  
0      97210  [NW 53rd Ave., Portland, OR 97210]  4cc8a97041e754818ff25784  
3        NaN                      [Portland, OR]  5532ac77498eca93a67815c2  
5        NaN                      [Portland, OR]  5176c925e4b056ce9504180d  
6        NaN                      [Portland, OR]  56ed8f45498ef3bb0236c48d  
7        NaN                      [Portland, OR]  521d8d2c11d2beb756541449

In [470]:
trail_dataframe_filtered.name

0                           Forest Park - Keil Trail
3                          Forest Park - Maple Trail
5                    Willamette River Greenway Trail
6                                SW Pedestrian Trail
7                                     Hillside Trail
8                       Peaks and Trails Outdoor Co.
11                         Trailhead coffee Roasters
12    Douglas Fir Portland State University Trailers
Name: name, dtype: object

# A Map of Apartments, Pet related services and goods, Parks & Trails, and Grocery stores

In [471]:
venues_map = folium.Map(width=900, height=800,location=[latitude, longitude], zoom_start=15) # generate map centred around central downtown with 2 mile / 3218.69 meter radius of Downtown

# add a red circle marker to represent the center point
folium.CircleMarker(
    [latitude, longitude],
    radius=10,
    color='red',
    popup='Central downtown',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# add the Apartments as circle markers
for lat, lng, label in zip(apartments_dataframe_filtered.lat, apartments_dataframe_filtered.lng, apartments_dataframe_filtered.categories):
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        color='magenta',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# add the Pet related services and stores as blue circle markers
for lat, lng, label in zip(pet_dataframe_filtered.lat, pet_dataframe_filtered.lng, pet_dataframe_filtered.categories):
    folium.CircleMarker(
        [lat, lng],
        radius=6,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)
    
# add the groceries as circle markers green and lime
for lat, lng, label in zip(grocery_dataframe_filtered.lat, grocery_dataframe_filtered.lng, grocery_dataframe_filtered.categories):
    folium.CircleMarker(
        [lat, lng],
        radius=4,
        color='green',
        popup=label,
        fill = True,
        fill_color='lime',
        fill_opacity=0.6
    ).add_to(venues_map)

# add the groceries as circle markers green and lime
for lat, lng, label in zip(trail_dataframe_filtered.lat, trail_dataframe_filtered.lng, trail_dataframe_filtered.categories):
    folium.CircleMarker(
        [lat, lng],
        radius=8,
        color='yellow',
        popup=label,
        fill = True,
        fill_color='brown',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map


# Analyzing Apartments, the priority criteria for *Paw & Claw*

In [472]:
# one hot encoding
apartments_onehot = pd.get_dummies(apartments_dataframe_filtered[['categories']], prefix="", prefix_sep="")

# add distance column back to dataframe
apartments_onehot['name'] = apartments_dataframe_filtered['name'] 

# move neighborhood column to the first column
fixed_columns = [apartments_onehot.columns[-1]] + list(apartments_onehot.columns[:-1])
apartments_onehot = apartments_onehot[fixed_columns]

apartments_onehot.head()

name  Music Venue  \
0  Bar Bar Apartment at Mississippi Studios            1   
1            10Th@Hoyt Apartment Think Tank            0   
2                    Essex House Apartments            0   
3                   Pacifc Tower Apartments            0   
4                       38 Davis Apartments            0   

   Residential Building (Apartment / Condo)  
0                                         0  
1                                         1  
2                                         1  
3                                         1  
4                                         1

In [473]:
apartments_onehot.shape

(24, 3)

In [474]:
apartments_grouped = apartments_onehot.groupby('name').mean().reset_index()
apartments_grouped.head()

name  Music Venue  \
0       10Th@Hoyt Apartment Think Tank            0   
1               10th @ Hoyt Apartments            0   
2                11 Marche' Apartments            0   
3                  38 Davis Apartments            0   
4  5th Ave Place Apartments -silverado            0   

   Residential Building (Apartment / Condo)  
0                                         1  
1                                         1  
2                                         1  
3                                         1  
4                                         1

In [475]:
apartments_grouped.shape

(24, 3)

In [476]:
num_top_venues = 10

for hood in apartments_grouped['name']:
    print("----"+hood+"----")
    temp = apartments_grouped[apartments_grouped['name'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----10Th@Hoyt Apartment Think Tank----
                                      venue  freq
0  Residential Building (Apartment / Condo)   1.0
1                               Music Venue   0.0


----10th @ Hoyt Apartments----
                                      venue  freq
0  Residential Building (Apartment / Condo)   1.0
1                               Music Venue   0.0


----11 Marche' Apartments----
                                      venue  freq
0  Residential Building (Apartment / Condo)   1.0
1                               Music Venue   0.0


----38 Davis Apartments----
                                      venue  freq
0  Residential Building (Apartment / Condo)   1.0
1                               Music Venue   0.0


----5th Ave Place Apartments -silverado----
                                      venue  freq
0  Residential Building (Apartment / Condo)   1.0
1                               Music Venue   0.0


----Alder St. Apartments----
                                      v

Sorting the venues

In [477]:
print('There are {} uniques categories.'.format(len(apartments_dataframe_filtered['categories'].unique())))

There are 2 uniques categories.


# Travel Agencies

In [478]:
search_query = 'Travel'
categoryId = '4d4b7105d754a06378d81259'
radius = 1000
print(search_query + ' .... OK!')

Travel .... OK!


In [479]:
# The code was removed by Watson Studio for sharing.

In [480]:
travel = requests.get(url).json()

In [481]:
# assign relevant part of JSON to venues
venues = travel['response']['venues']
# tranform venues into a dataframe
travel_dataframe = json_normalize(venues)
travel_dataframe.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:4: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  after removing the cwd from sys.path.


id                            name  \
0  50dc99dd582fc4391173affc  Travel Portland Visitor Center   
1  4bedb0c1767dc9b6f751d3e9                  Azumano Travel   
2  505758bde4b04df2359d0475                   Travel Oregon   
3  5a2af05695da0c767ab3a5db        Travel Visa Pro Portland   
4  4b8ef785f964a520bd4133e3                 Travel Portland   

                                          categories    referralId  hasPerk  \
0  [{'id': '4f4530164b9074f6e4fb00ff', 'name': 'T...  v-1612779248    False   
1                                                 []  v-1612779248    False   
2  [{'id': '50328a8e91d4c4b30a586d6c', 'name': 'N...  v-1612779248    False   
3  [{'id': '56aa371be4b08b9a8d57352f', 'name': 'A...  v-1612779248    False   
4  [{'id': '4f04b08c2fb6e1c99f3db0bd', 'name': 'T...  v-1612779248    False   

          location.address location.crossStreet  location.lat  location.lng  \
0         877 SW Taylor St        at SW 9th Ave     45.518461   -122.681491   
1           400 SW 4th Ave       at SW Stark St     45.520531   -122.675149   
2  317 SW Alder St Ste 200                  NaN     45.519111   -122.675029   
3   121 SW Salmon St Fl 11                  NaN     45.523713   -122.676677   
4         1000 SW Broadway                 Main     45.516933   -122.680573   

                             location.labeledLatLngs  location.distance  \
0  [{'label': 'display', 'lat': 45.5184607, 'lng'...                602   
1  [{'label': 'display', 'lat': 45.52053133778469...                 80   
2  [{'label': 'display', 'lat': 45.51911089901961...                142   
3  [{'label': 'display', 'lat': 45.52371332594481...                431   
4  [{'label': 'display', 'lat': 45.516933, 'lng':...                619   

  location.postalCode location.cc location.city location.state  \
0               97205          US      Portland             OR   
1               97204          US      Portland             OR   
2               97204          US      Portland             OR   
3               97204          US      Portland             OR   
4               97205          US      Portland             OR   

  location.country                          location.formattedAddress  \
0    United States  [877 SW Taylor St (at SW 9th Ave), Portland, O...   
1    United States  [400 SW 4th Ave (at SW Stark St), Portland, OR...   
2    United States      [317 SW Alder St Ste 200, Portland, OR 97204]   
3    United States       [121 SW Salmon St Fl 11, Portland, OR 97204]   
4    United States      [1000 SW Broadway (Main), Portland, OR 97205]   

  venuePage.id  
0     44268773  
1          NaN  
2          NaN  
3          NaN  
4          NaN

In [482]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in travel_dataframe.columns if col.startswith('location.')] + ['id']
travel_dataframe_filtered = travel_dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
travel_dataframe_filtered['categories'] = travel_dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
travel_dataframe_filtered.columns = [column.split('.')[-1] for column in travel_dataframe_filtered.columns]

travel_dataframe_filtered.head(50)

name                  categories  \
0   Travel Portland Visitor Center  Tourist Information Center   
1                   Azumano Travel                        None   
2                    Travel Oregon                  Non-Profit   
3         Travel Visa Pro Portland             Airport Service   
4                  Travel Portland               Travel Agency   
5  Willamette International Travel              General Travel   
6       Travelex Currency Services           Currency Exchange   
7       Travelex Currency Services           Currency Exchange   
8      AAA Portland Service Center               Travel Agency   

                   address     crossStreet        lat         lng  \
0         877 SW Taylor St   at SW 9th Ave  45.518461 -122.681491   
1           400 SW 4th Ave  at SW Stark St  45.520531 -122.675149   
2  317 SW Alder St Ste 200             NaN  45.519111 -122.675029   
3   121 SW Salmon St Fl 11             NaN  45.523713 -122.676677   
4         1000 SW Broadway            Main  45.516933 -122.680573   
5        1314 NW Irving St             NaN  45.527696 -122.684846   
6        321 SW 6th Avenue             NaN  45.521500 -122.677902   
7           900 SW 6th Ave             NaN  45.517368 -122.679650   
8         600 SW Market St             NaN  45.512461 -122.682387   

                                      labeledLatLngs  distance postalCode  cc  \
0  [{'label': 'display', 'lat': 45.5184607, 'lng'...       602      97205  US   
1  [{'label': 'display', 'lat': 45.52053133778469...        80      97204  US   
2  [{'label': 'display', 'lat': 45.51911089901961...       142      97204  US   
3  [{'label': 'display', 'lat': 45.52371332594481...       431      97204  US   
4  [{'label': 'display', 'lat': 45.516933, 'lng':...       619      97205  US   
5  [{'label': 'display', 'lat': 45.527696, 'lng':...      1173      97209  US   
6  [{'label': 'display', 'lat': 45.52149963378906...       321      97204  US   
7  [{'label': 'display', 'lat': 45.51736829551206...       532      97204  US   
8  [{'label': 'display', 'lat': 45.51246129877682...      1076      97201  US   

       city state        country  \
0  Portland    OR  United States   
1  Portland    OR  United States   
2  Portland    OR  United States   
3  Portland    OR  United States   
4  Portland    OR  United States   
5  Portland    OR  United States   
6  Portland    OR  United States   
7  Portland    OR  United States   
8  Portland    OR  United States   

                                    formattedAddress                        id  
0  [877 SW Taylor St (at SW 9th Ave), Portland, O...  50dc99dd582fc4391173affc  
1  [400 SW 4th Ave (at SW Stark St), Portland, OR...  4bedb0c1767dc9b6f751d3e9  
2      [317 SW Alder St Ste 200, Portland, OR 97204]  505758bde4b04df2359d0475  
3       [121 SW Salmon St Fl 11, Portland, OR 97204]  5a2af05695da0c767ab3a5db  
4      [1000 SW Broadway (Main), Portland, OR 97205]  4b8ef785f964a520bd4133e3  
5            [1314 NW Irving St, Portland, OR 97209]  4e4568ec45dd2955e5894418  
6            [321 SW 6th Avenue, Portland, OR 97204]  4ecd6140aad2f45d5032fa05  
7               [900 SW 6th Ave, Portland, OR 97204]  4ecd6140aad2f45d5032fa04  
8             [600 SW Market St, Portland, OR 97201]  4c506ea38edf0f47c50747af

There are a few Categories with None values, I am replacing with General Travel

In [483]:
travel_dataframe_filtered.categories.fillna(value='General Travel', inplace=True)

In [484]:
travel_dataframe_filtered.drop(["crossStreet", "cc", "city", "state","country","crossStreet"], axis=1, inplace = True)

In [485]:
travel_dataframe_filtered.head(10)

name                  categories  \
0   Travel Portland Visitor Center  Tourist Information Center   
1                   Azumano Travel              General Travel   
2                    Travel Oregon                  Non-Profit   
3         Travel Visa Pro Portland             Airport Service   
4                  Travel Portland               Travel Agency   
5  Willamette International Travel              General Travel   
6       Travelex Currency Services           Currency Exchange   
7       Travelex Currency Services           Currency Exchange   
8      AAA Portland Service Center               Travel Agency   

                   address        lat         lng  \
0         877 SW Taylor St  45.518461 -122.681491   
1           400 SW 4th Ave  45.520531 -122.675149   
2  317 SW Alder St Ste 200  45.519111 -122.675029   
3   121 SW Salmon St Fl 11  45.523713 -122.676677   
4         1000 SW Broadway  45.516933 -122.680573   
5        1314 NW Irving St  45.527696 -122.684846   
6        321 SW 6th Avenue  45.521500 -122.677902   
7           900 SW 6th Ave  45.517368 -122.679650   
8         600 SW Market St  45.512461 -122.682387   

                                      labeledLatLngs  distance postalCode  \
0  [{'label': 'display', 'lat': 45.5184607, 'lng'...       602      97205   
1  [{'label': 'display', 'lat': 45.52053133778469...        80      97204   
2  [{'label': 'display', 'lat': 45.51911089901961...       142      97204   
3  [{'label': 'display', 'lat': 45.52371332594481...       431      97204   
4  [{'label': 'display', 'lat': 45.516933, 'lng':...       619      97205   
5  [{'label': 'display', 'lat': 45.527696, 'lng':...      1173      97209   
6  [{'label': 'display', 'lat': 45.52149963378906...       321      97204   
7  [{'label': 'display', 'lat': 45.51736829551206...       532      97204   
8  [{'label': 'display', 'lat': 45.51246129877682...      1076      97201   

                                    formattedAddress                        id  
0  [877 SW Taylor St (at SW 9th Ave), Portland, O...  50dc99dd582fc4391173affc  
1  [400 SW 4th Ave (at SW Stark St), Portland, OR...  4bedb0c1767dc9b6f751d3e9  
2      [317 SW Alder St Ste 200, Portland, OR 97204]  505758bde4b04df2359d0475  
3       [121 SW Salmon St Fl 11, Portland, OR 97204]  5a2af05695da0c767ab3a5db  
4      [1000 SW Broadway (Main), Portland, OR 97205]  4b8ef785f964a520bd4133e3  
5            [1314 NW Irving St, Portland, OR 97209]  4e4568ec45dd2955e5894418  
6            [321 SW 6th Avenue, Portland, OR 97204]  4ecd6140aad2f45d5032fa05  
7               [900 SW 6th Ave, Portland, OR 97204]  4ecd6140aad2f45d5032fa04  
8             [600 SW Market St, Portland, OR 97201]  4c506ea38edf0f47c50747af

In [486]:
# create map of Tracel using latitude and longitude values
travel_map = folium.Map(width=1000, height=800,location=[latitude, longitude], zoom_start=15) # generate map centred around central downtown with 2 mile / 3218.69 meter radius of Downtown
# add markers to map
for lat, lng, label in zip(travel_dataframe_filtered.lat, travel_dataframe_filtered.lng, travel_dataframe_filtered.categories):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=7,
        popup=label,
        color='black',
        fill=True,
        fill_color='orange',
        fill_opacity=0.7,
        parse_html=False).add_to(travel_map)  
    
travel_map


Sorting

In [487]:
travel_grouped = travel_onehot.groupby('name').mean().reset_index()
travel_grouped

name  Airport Service  Currency Exchange  \
0      AAA Portland Service Center                0                  0   
1                   Azumano Travel                0                  0   
2                    Travel Oregon                0                  0   
3                  Travel Portland                0                  0   
4   Travel Portland Visitor Center                0                  0   
5         Travel Visa Pro Portland                1                  0   
6       Travelex Currency Services                0                  1   
7  Willamette International Travel                0                  0   

   General Travel  Non-Profit  Tourist Information Center  Travel Agency  
0               0           0                           0              1  
1               1           0                           0              0  
2               0           1                           0              0  
3               0           0                           0              1  
4               0           0                           1              0  
5               0           0                           0              0  
6               0           0                           0              0  
7               1           0                           0              0

In [488]:
travel_grouped.shape

(8, 7)

Top most common venues

In [489]:
num_top_venues = 5

for hood in travel_grouped['name']:
    print("----"+hood+"----")
    temp = travel_grouped[travel_grouped['name'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----AAA Portland Service Center----
               venue  freq
0      Travel Agency   1.0
1    Airport Service   0.0
2  Currency Exchange   0.0
3     General Travel   0.0
4         Non-Profit   0.0


----Azumano Travel----
                        venue  freq
0              General Travel   1.0
1             Airport Service   0.0
2           Currency Exchange   0.0
3                  Non-Profit   0.0
4  Tourist Information Center   0.0


----Travel Oregon----
                        venue  freq
0                  Non-Profit   1.0
1             Airport Service   0.0
2           Currency Exchange   0.0
3              General Travel   0.0
4  Tourist Information Center   0.0


----Travel Portland----
               venue  freq
0      Travel Agency   1.0
1    Airport Service   0.0
2  Currency Exchange   0.0
3     General Travel   0.0
4         Non-Profit   0.0


----Travel Portland Visitor Center----
                        venue  freq
0  Tourist Information Center   1.0
1             Airpor

In [490]:
travel_grouped.shape

(8, 7)

In [491]:
all(isinstance(column, str) for column in travel_dataframe_filtered.columns)

True

In [492]:
travel_dataframe_filtered['Total'] = travel_dataframe_filtered.sum(axis=1)

# let's view the first five elements and see how the dataframe was changed
travel_dataframe_filtered.head()

name                  categories  \
0  Travel Portland Visitor Center  Tourist Information Center   
1                  Azumano Travel              General Travel   
2                   Travel Oregon                  Non-Profit   
3        Travel Visa Pro Portland             Airport Service   
4                 Travel Portland               Travel Agency   

                   address        lat         lng  \
0         877 SW Taylor St  45.518461 -122.681491   
1           400 SW 4th Ave  45.520531 -122.675149   
2  317 SW Alder St Ste 200  45.519111 -122.675029   
3   121 SW Salmon St Fl 11  45.523713 -122.676677   
4         1000 SW Broadway  45.516933 -122.680573   

                                      labeledLatLngs  distance postalCode  \
0  [{'label': 'display', 'lat': 45.5184607, 'lng'...       602      97205   
1  [{'label': 'display', 'lat': 45.52053133778469...        80      97204   
2  [{'label': 'display', 'lat': 45.51911089901961...       142      97204   
3  [{'label': 'display', 'lat': 45.52371332594481...       431      97204   
4  [{'label': 'display', 'lat': 45.516933, 'lng':...       619      97205   

                                    formattedAddress  \
0  [877 SW Taylor St (at SW 9th Ave), Portland, O...   
1  [400 SW 4th Ave (at SW Stark St), Portland, OR...   
2      [317 SW Alder St Ste 200, Portland, OR 97204]   
3       [121 SW Salmon St Fl 11, Portland, OR 97204]   
4      [1000 SW Broadway (Main), Portland, OR 97205]   

                         id       Total  
0  50dc99dd582fc4391173affc  524.836970  
1  4bedb0c1767dc9b6f751d3e9    2.845382  
2  505758bde4b04df2359d0475   64.844082  
3  5a2af05695da0c767ab3a5db  353.847037  
4  4b8ef785f964a520bd4133e3  541.836360

In [493]:
print ('data dimensions:', travel_dataframe_filtered.shape)

data dimensions: (9, 11)


Sort by distance closest to central downtown

In [494]:
travel_dataframe_filtered.sort_values(by=['distance'], inplace=True)

In [495]:
travel_dataframe_filtered.head(10)

name                  categories  \
1                   Azumano Travel              General Travel   
2                    Travel Oregon                  Non-Profit   
6       Travelex Currency Services           Currency Exchange   
3         Travel Visa Pro Portland             Airport Service   
7       Travelex Currency Services           Currency Exchange   
0   Travel Portland Visitor Center  Tourist Information Center   
4                  Travel Portland               Travel Agency   
8      AAA Portland Service Center               Travel Agency   
5  Willamette International Travel              General Travel   

                   address        lat         lng  \
1           400 SW 4th Ave  45.520531 -122.675149   
2  317 SW Alder St Ste 200  45.519111 -122.675029   
6        321 SW 6th Avenue  45.521500 -122.677902   
3   121 SW Salmon St Fl 11  45.523713 -122.676677   
7           900 SW 6th Ave  45.517368 -122.679650   
0         877 SW Taylor St  45.518461 -122.681491   
4         1000 SW Broadway  45.516933 -122.680573   
8         600 SW Market St  45.512461 -122.682387   
5        1314 NW Irving St  45.527696 -122.684846   

                                      labeledLatLngs  distance postalCode  \
1  [{'label': 'display', 'lat': 45.52053133778469...        80      97204   
2  [{'label': 'display', 'lat': 45.51911089901961...       142      97204   
6  [{'label': 'display', 'lat': 45.52149963378906...       321      97204   
3  [{'label': 'display', 'lat': 45.52371332594481...       431      97204   
7  [{'label': 'display', 'lat': 45.51736829551206...       532      97204   
0  [{'label': 'display', 'lat': 45.5184607, 'lng'...       602      97205   
4  [{'label': 'display', 'lat': 45.516933, 'lng':...       619      97205   
8  [{'label': 'display', 'lat': 45.51246129877682...      1076      97201   
5  [{'label': 'display', 'lat': 45.527696, 'lng':...      1173      97209   

                                    formattedAddress  \
1  [400 SW 4th Ave (at SW Stark St), Portland, OR...   
2      [317 SW Alder St Ste 200, Portland, OR 97204]   
6            [321 SW 6th Avenue, Portland, OR 97204]   
3       [121 SW Salmon St Fl 11, Portland, OR 97204]   
7               [900 SW 6th Ave, Portland, OR 97204]   
0  [877 SW Taylor St (at SW 9th Ave), Portland, O...   
4      [1000 SW Broadway (Main), Portland, OR 97205]   
8             [600 SW Market St, Portland, OR 97201]   
5            [1314 NW Irving St, Portland, OR 97209]   

                         id        Total  
1  4bedb0c1767dc9b6f751d3e9     2.845382  
2  505758bde4b04df2359d0475    64.844082  
6  4ecd6140aad2f45d5032fa05   243.843597  
3  5a2af05695da0c767ab3a5db   353.847037  
7  4ecd6140aad2f45d5032fa04   454.837718  
0  50dc99dd582fc4391173affc   524.836970  
4  4b8ef785f964a520bd4133e3   541.836360  
8  4c506ea38edf0f47c50747af   998.830075  
5  4e4568ec45dd2955e5894418  1095.842850

# Results and Discussion

Our analysis shows that there is high density of apartments and condos around in the Portland University are to the west of the city center point likely due to a student and faculty population.  Additionally there are trails and parks nearby to walk pets.  There are several smaller groceries and convenience stores north of campus, with larger stores central downtown.  There are travel related services present that serve the student population and nearby downtown.  There are no other Pet Services potentially competing with Paw & Claw in the one mile radius, according to Foursquare's data.

A challenge for *Paw & Claw* would be a lack of available real estate.  A web search for available commercial propertly turned up few results.

I had wished to find tips for Travel Agencies, but no results were found from Foursquare.



# Conclusion

**What is the ideal location intersecting near pet supply stores, and parks, with current limited existing Pet Services?** 

I reviewed the map points to determine ideal location zip codes 97201,97205, 97204, 97209 are the best prospects. There is a dog park and dog run near the waterfront just east of campus in zip code 97201 and this is the optimimum location, with two small parks in between and a major trail along the waterfront.

There are 8 travel agencies, and specialty services such as passport, currency exchange, and hip travel hubs like a Segway rental and ride shares, to support an international population.  I highly recommend this area for Paw & Claw's expansion plans with consideration to real estate availability and planning. The density of travel agencies within 1 radius mile is dense and Paw & Clas should consider this a strong indicator on whether to add a boarding service to their new location.